### ISSUE TITLE
Error Handling section is sloppy with terms

### ISSUE THREAD

**Issue by martinduke** [ietf-wg-masque/draft-ietf-masque-h3-datagram#200]
In Sec 3.3:

When an error occurs in processing the Capsule Protocol, the receiver
   MUST treat the message as malformed or incomplete...

there are 2 other reference to malformed "messages" as a result of Capsule shenanigans.

but in the introduction it says that datagrams "are not part of message content". 

So what is "the message" that should be treated this way? I think you mean the Request or Response that initiated the Capsule protocol.

If that's correct, a more precise phrasing might be "MUST act as if the message that initiated the capsule protocol was malformed".

**Comment by LPardue** [comment]
Agree we can probably tighten things up here. 

For one, the definition of "data stream" in 3.1 presently says

> This specification defines the "data stream" of an HTTP request as the bidirectional stream of bytes that follows the header section of the request message and the final, successful (i.e., 2xx) response message.
> 
> In HTTP/1.x, the data stream consists of all bytes on the connection that follow the blank line that concludes either the request header section, or the response header section. As a result, only a single HTTP request starting the capsule protocol can be sent on HTTP/1.x connections.
> 
> In HTTP/2 and HTTP/3, the data stream of a given HTTP request consists of all bytes sent in DATA frames with the corresponding stream ID.

In isolation, this statement leaves it a little unclear whether trailing headers can appear after the data stream. In [7540bis section 8.5](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#section-8.5), that defines how CONNECT works, things are a lot clearer:

> Frame types other than [DATA](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#DATA) or stream management frames ([RST_STREAM](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#RST_STREAM), [WINDOW_UPDATE](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#WINDOW_UPDATE), and [PRIORITY](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#PRIORITY)) MUST NOT be sent on a connected stream and MUST be treated as a [stream error](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#StreamErrorHandler) ([Section 5.4.2](https://httpwg.org/http2-spec/draft-ietf-httpbis-http2bis.html#StreamErrorHandler)) if received.

My working assumption has been that data streams have similar requirements, but we don't explicitly state that anywhere, which is not great. If we tighten that up then it should become clearer that, once the capsule protocol is successfully negotiated, the HTTP message ended after the first HEADERS frames. We would still expect endpoints to apply some rules to frame processing on these data streams, and if there was a problem the data stream is malformed. Today we equate malformed data streams to malformed requests, and therefore delegate the handling of such problems (reset stream + error code) to the respect section in H2 or H3. I wonder actually if we should just state some explicit stream termination requirements. This does not change the actions that endpoint might take, but it does clarify the circumstances under which they should happen.

I'll prepare a PR for this to help illustrate my thinking.

To summarise:

1) If an endpoint screws up meeting the requirements of request or response field sections that bootstrap a capsule protocol instance, that is clearly a standard "malformed message"
2) If, after a capsule protocol instance has successully been created, an endpoint screws up anything on the data stream, that does not seem like a "malformed message"

**Comment by martinduke** [comment]
I can understand the impulse to avoid the term "stream" because then you're in the business of defining the behavior per HTTP version, but I'm sure you'll come up with some good language.

**Comment by DavidSchinazi** [comment]
+1 to what Lucas said above
